# Imports

In [1]:
# Print every output from a cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# EDA
import numpy as np
import pandas as pd

# Visualizations
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

# Plot pandas objects using plotly.
pd.options.plotting.backend = "plotly"

import cufflinks as cf
init_notebook_mode(connected=True)
cf.go_offline()

# Data Source

### Data sourced from The Bureau of Labor Statistics public database ( LABSTAT ): https://download.bls.gov
<br> BLS Consumer Price Index Data 2022: https://download.bls.gov/pub/time.series/cu/

# Metadata

### CPI-U Description: 

The Consumer Price Index (CPI) All Urban Consumers (CU) is a statistical measure
of change, over time, of the prices of goods and services in major 
expenditure groups--such as food, housing, apparel, transportation, and 
medical care--typically purchased by urban consumers. Essentially, it 
compares the cost of a sample "market basket" of goods and services in a 
specific month relative to the cost of the same "market basket" in an 
earlier reference period. This reference period is designated as the base 
period.

### CPI-U Database Files: 

    cu.data.0.Current			-  All current year-to-date 
	cu.data.1.AllItems			-  All items (item_code AA0, SA0)
	cu.data.2.Summaries			-  Summaries (item_code SA0, SAF,
						   SAH, SAA, SAT, SAM, SAR, SAE, SAG,
						   SAS, SAC)
	cu.data.3.AsizeNorthEast		-  A-Size areas in Northeast
						   (area_code A1 ...)
	cu.data.4.AsizeNorthCentral		-  A-size areas in North Central
						   (area_code A2 ...)
	cu.data.5.AsizeSouth			-  A-Size areas in South
						   (area_code A3 ...)
	cu.data.6.AsizeWest			-  A-Size areas in West
						   (area_code A4 ...)
	cu.data.7.OtherNorthEast		-  All other Northeast
						   (area_code 01, X1, D1)
	cu.data.8.OtherNorthCentral		-  All other North Central
						   (area_code 02, X2, D2)
	cu.data.9.OtherSouth			-  All other in South
						   (area_code 03, X3, D3)
	cu.data.10.OtherWest			-  All other in West
						   (area_code 04, X4)  
	cu.data.11.USFoodBeverage		-  All US Food and Beverage
						   (area_code 0000, item_code SAF, SEF)
	cu.data.12.USHousing			-  All US Housing (area_code 0000, 
						   item_code SAH, SEH)
	cu.data.13.USApparel			-  All US Apparel (area_code 0000,
						   item_code SAA, SEA)
	cu.data.14.USTransportation		-  All US Transportation (area_code
						   0000, item_code SAT, SET)
	cu.data.15.USMedical			-  All US Medical (area_code 0000, 
						   item_code SAM, SEM, SS57)
	cu.data.16.USRecreation			-  All US Recreation (area_code 0000,
						   item_code SAR, SER, SS31, SS61, SS62)
	cu.data.17.USEducationAndCommunication	-  All US Education and Communication
						   (area_code 0000, item_code SAE,
						    SEE, SS27)
	cu.data.18.USOtherGoodsAndServices	-  All US Other Goods and Services
						   (area_code 0000, item_code SAG, SEG;
						    SS33)
	cu.data.19.PopulationSize		-  All Population-size (area_code
						   A000, X000, D000)
	cu.data.20.USCommoditiesServicesSpecial	-  All US Commodity and Services and 
						   Special(area_code 0000, item_code 
						   SA0, SAC, SAN, SAS)
	cu.area					-  Area codes		mapping file
	cu.contacts				-  Contacts for cu survey 
	cu.footnote				-  Footnote codes	mapping file
	cu.item					-  Item codes		mapping file
	cu.MapErrors (TBR)			-  Map error codes	mapping file
	cu.period				-  Period codes 	mapping file
	cu.series				-  All series and their beginning and end dates
	cu.txt					-  General information

### CPI-U Database Elements:

    Data Element	Length		Value(Example)			Description

    area_code	4		N or S, ,000-999		Unique code used to identify
				Ex: A100 		a specific geographic area.
					
    area_name	80		Text 			Name of specific geographic
				Ex: MIAMI, FLA		area.
				
    base_code	1		S=Standard reference 	Code identifying the type of
				base			base period used in index.
				A=Alternate reference 
				base	

    base_period	20		Identifies the base  	The actual base period used
				period			in calculating the index.
				Ex: 1982-84=100	
				     
    begin_period	3		M01-M13 or S01-S03	Identifies first data observation
				Ex: MO6=June		within the first year for which 
				(M=Monthly, M13=Annual	data is available for a given
				Avg, S=Semi-Annually)	time series.
				
    begin_year	4		YYYY			Identifies first year for which
				Ex: 1975		data is available for a given
							time series.

    end_period	3		M01-M13 or S01-S03	Identifies last data observation
				Ex: M06=June		within the last year for which 
				(M=Monthly, M13=Annual	data is available for a given
				Avg, S=Semi-Annually)	time series.
				
    end_year	4		YYYY			Identifies last year for which 
				Ex: 1980		data is available for a given  
							time series.

    footnote_code	1		R			Identifies footnote for the data 
							series.

    footnote_text	100		Text			Contains the text of the footnote.

    item_code	8		SA0E			Identifies item for which 
							data observations pertain.

    item_name	100		Text			Full names of items.
				
    period_abbr	5		Text	 		Abbreviation of period name.
				Ex: JUN

    period		3		M01-M13 or S01-S03	Identifies period for which
				Ex: M06=June		data is observed. 
				(M=Monthly, M13=Annual	
				Avg, S=Semi-Annually)
				
    period_name	20		Text		 	Full name of period to which
				Ex: June		the data observation refers.
				
    periodicity_	1		S=Semi-Annual		Frequency of data
    code				R=Regular		observation.

    seasonal	1		S=Seasonally		Code identifying whether the
				  Adjusted 		data are seasonally adjusted.
				U=Unadjusted			  		

    series_id	17		CUSR0000SA0		Code identifying the specific 
							series.

    value		12		53.3			Price index for item.
	   
    year		4		YYYY			Identifies year of observation.
				Ex: 1990	

# Data Acquisition

### For the latest data, pull from LABSTAT the data file ( data.0.Current ) and the significant mapping files ( area, item, period, series, seasonal; footnote file is empty ):

In [2]:
# Define URL strings.
url = 'https://download.bls.gov/pub/time.series/'
cu = 'cu/cu.'

# Define BLS API function.
def blsAPI(string):
    return(pd.read_csv(url+cu+string,sep='\t'))

# Example dataframe for pulling cu.seasonal file from the BLS website:
example_df = blsAPI('seasonal')
example_df.head()

seasonal_code            seasonal_text
0             S      Seasonally Adjusted
1             U  Not Seasonally Adjusted

## CPI Data

### For the latest data, pull from LABSTAT the data file ( data.0.Current ) and the significant mapping files ( area, item, period, seasonal, series; footnote file is empty ):

In [3]:
cpi_df = blsAPI('data.0.Current')
area_df = blsAPI('area')
item_df = blsAPI('item')
period_df = blsAPI('period')
seasonal_df = blsAPI('seasonal')
series_df = blsAPI('series')

# EDA

## area_df

In [4]:
area_df.head()
area_df.info()

area_code          area_name  display_level selectable  sort_sequence
0      0000  U.S. city average              0          T              1
1      0100          Northeast              0          T              5
2      0110        New England              1          T             10
3      0120    Middle Atlantic              1          T             11
4      0200            Midwest              0          T             14

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58 entries, 0 to 57
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   area_code      58 non-null     object
 1   area_name      58 non-null     object
 2   display_level  58 non-null     int64 
 3   selectable     58 non-null     object
 4   sort_sequence  58 non-null     int64 
dtypes: int64(2), object(3)
memory usage: 2.4+ KB


### All of the area codes are 4 digits long.

In [5]:
area_df['area_code'].apply(len).value_counts()

4    58
Name: area_code, dtype: int64

### There are 58 unique area codes, each corresponding to some region ( the meaning of some, like 'West - Size Class A', are not readily apparent; size class may have to do with population ).

In [6]:
area_df.nunique()
area_df['area_name'].unique()

area_code        58
area_name        58
display_level     2
selectable        1
sort_sequence    58
dtype: int64

array(['U.S. city average', 'Northeast', 'New England', 'Middle Atlantic',
       'Midwest', 'East North Central', 'West North Central', 'South',
       'South Atlantic', 'East South Central', 'West South Central',
       'West', 'Mountain', 'Pacific', 'Pittsburgh, PA',
       'Cleveland-Akron, OH', 'Milwaukee-Racine, WI',
       'Cincinnati-Hamilton, OH-KY-IN', 'Kansas City, MO-KS',
       'Washington-Baltimore, DC-MD-VA-WV',
       'Los Angeles-Riverside-Orange County, CA', 'Portland-Salem, OR-WA',
       'Size Class D', 'Midwest - Size Class D', 'South - Size Class D',
       'Size Class B/C', 'Northeast - Size Class B/C',
       'Midwest - Size Class B/C', 'South - Size Class B/C',
       'West - Size Class B/C', 'Size Class A',
       'Northeast - Size Class A', 'Boston-Cambridge-Newton, MA-NH',
       'New York-Newark-Jersey City, NY-NJ-PA',
       'Philadelphia-Camden-Wilmington, PA-NJ-DE-MD',
       'Midwest - Size Class A', 'Chicago-Naperville-Elgin, IL-IN-WI',
       'Detroit

## cpi_df

In [7]:
cpi_df.head()
cpi_df.info()

series_id          year period         value  footnote_codes
0  CUSR0000SA0        1997    M01         159.4             NaN
1  CUSR0000SA0        1997    M02         159.7             NaN
2  CUSR0000SA0        1997    M03         159.8             NaN
3  CUSR0000SA0        1997    M04         159.9             NaN
4  CUSR0000SA0        1997    M05         159.9             NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 960933 entries, 0 to 960932
Data columns (total 5 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   series_id          960933 non-null  object 
 1   year               960933 non-null  int64  
 2   period             960933 non-null  object 
 3          value       960933 non-null  float64
 4   footnote_codes     0 non-null       float64
dtypes: float64(2), int64(1), object(2)
memory usage: 36.7+ MB


### The series_id and value feature names contain tabs; let's remove them.

In [8]:
cpi_df.columns
cpi_df = cpi_df.rename(columns={'series_id        ':'series_id', '       value':'value'})
cpi_df.columns

Index(['series_id        ', 'year', 'period', '       value',
       'footnote_codes'],
      dtype='object')

Index(['series_id', 'year', 'period', 'value', 'footnote_codes'], dtype='object')

### The footnote_codes values are all null; let's drop this feature.

In [9]:
cpi_df.head()
cpi_df = cpi_df.drop(['footnote_codes'],axis=1)
cpi_df.head()

series_id  year period  value  footnote_codes
0  CUSR0000SA0        1997    M01  159.4             NaN
1  CUSR0000SA0        1997    M02  159.7             NaN
2  CUSR0000SA0        1997    M03  159.8             NaN
3  CUSR0000SA0        1997    M04  159.9             NaN
4  CUSR0000SA0        1997    M05  159.9             NaN

series_id  year period  value
0  CUSR0000SA0        1997    M01  159.4
1  CUSR0000SA0        1997    M02  159.7
2  CUSR0000SA0        1997    M03  159.8
3  CUSR0000SA0        1997    M04  159.9
4  CUSR0000SA0        1997    M05  159.9

## item_df

In [10]:
item_df.head()
item_df.info()

item_code                                          item_name  display_level  \
0       AA0                               All items - old base              0   
1      AA0R  Purchasing power of the consumer dollar - old ...              0   
2       SA0                                          All items              0   
3      SA0E                                             Energy              1   
4     SA0L1                                All items less food              1   

  selectable  sort_sequence  
0          T              2  
1          T            400  
2          T              1  
3          T            375  
4          T            359

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   item_code      400 non-null    object
 1   item_name      400 non-null    object
 2   display_level  400 non-null    int64 
 3   selectable     400 non-null    object
 4   sort_sequence  400 non-null    int64 
dtypes: int64(2), object(3)
memory usage: 15.8+ KB


### There are 400 unique item_code values, each corresponding to some good or service specified in the item_name feature.

In [11]:
item_df.nunique()
item_df['item_name'].unique()

item_code        400
item_name        400
display_level      9
selectable         1
sort_sequence    400
dtype: int64

array(['All items - old base',
       'Purchasing power of the consumer dollar - old base', 'All items',
       'Energy', 'All items less food', 'All items less food and shelter',
       'All items less food, shelter, and energy',
       'All items less food, shelter, energy, and used cars and trucks',
       'All items less food and energy', 'All items less shelter',
       'All items less medical care', 'All items less energy',
       'Purchasing power of the consumer dollar', 'Apparel less footwear',
       'Apparel', "Men's and boys' apparel", "Women's and girls' apparel",
       'Commodities', 'Energy commodities', 'Commodities less food',
       'Commodities less food and beverages',
       'Commodities less food and energy commodities',
       'Commodities less food, energy, and used cars and trucks',
       'Durables', 'Education and communication', 'Education',
       'Communication', 'Information and information processing',
       'Education and communication commodities',
 

## period_df

### There are 16 unique periods, 12 corresponding to the months of the year, and 4 corresponding to an annual average ( one averaged over 12 months, other averaged over 2 half-years; should be equal? ).

In [12]:
period_df

period period_abbr     period_name
0     M01         JAN         January
1     M02         FEB        February
2     M03         MAR           March
3     M04         APR           April
4     M05         MAY             May
5     M06         JUN            June
6     M07         JUL            July
7     M08         AUG          August
8     M09         SEP       September
9     M10         OCT         October
10    M11         NOV        November
11    M12         DEC        December
12    M13       AN AV  Annual Average
13    S01       HALF1      First Half
14    S02       HALF2     Second Half
15    S03       AN AV  Annual Average

## seasonal_df

### The seasonal_df values are binary.

In [13]:
seasonal_df

seasonal_code            seasonal_text
0             S      Seasonally Adjusted
1             U  Not Seasonally Adjusted

In [14]:
# # Encode as dummy variables.
# pd.get_dummies(seasonal_df['seasonal_code'])

## series_df

In [15]:
series_df.head()
series_df.info()

series_id         area_code item_code seasonal periodicity_code base_code  \
0  CUSR0000SA0            0000       SA0        S                R         S   
1  CUSR0000SA0E           0000      SA0E        S                R         S   
2  CUSR0000SA0L1          0000     SA0L1        S                R         S   
3  CUSR0000SA0L12         0000    SA0L12        S                R         S   
4  CUSR0000SA0L12E        0000   SA0L12E        S                R         S   

   base_period                                       series_title  \
0  1982-84=100  All items in U.S. city average, all urban cons...   
1  1982-84=100  Energy in U.S. city average, all urban consume...   
2  1982-84=100  All items less food in U.S. city average, all ...   
3  1982-84=100  All items less food and shelter in U.S. city a...   
4  1982-84=100  All items less food, shelter, and energy in U....   

   footnote_codes  begin_year begin_period  end_year end_period  
0             NaN        1947          M01      2022        M06  
1             NaN        1957          M01      2022        M06  
2             NaN        1947          M01      2022        M06  
3             NaN        1967          M01      2022        M06  
4             NaN        1967          M01      2022        M06

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8096 entries, 0 to 8095
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   series_id          8096 non-null   object 
 1   area_code          8096 non-null   object 
 2   item_code          8096 non-null   object 
 3   seasonal           8096 non-null   object 
 4   periodicity_code   8096 non-null   object 
 5   base_code          8096 non-null   object 
 6   base_period        8096 non-null   object 
 7   series_title       8096 non-null   object 
 8   footnote_codes     0 non-null      float64
 9   begin_year         8096 non-null   int64  
 10  begin_period       8096 non-null   object 
 11  end_year           8096 non-null   int64  
 12  end_period         8096 non-null   object 
dtypes: float64(1), int64(2), object(10)
memory usage: 822.4+ KB


### The series_id feature name contains a tab whitespace; let's remove it.

In [16]:
series_df.columns
series_df = series_df.rename(columns={'series_id        ':'series_id'})
series_df.columns

Index(['series_id        ', 'area_code', 'item_code', 'seasonal',
       'periodicity_code', 'base_code', 'base_period', 'series_title',
       'footnote_codes', 'begin_year', 'begin_period', 'end_year',
       'end_period'],
      dtype='object')

Index(['series_id', 'area_code', 'item_code', 'seasonal', 'periodicity_code',
       'base_code', 'base_period', 'series_title', 'footnote_codes',
       'begin_year', 'begin_period', 'end_year', 'end_period'],
      dtype='object')

### There are 8096 unique series_id values, but only 4229 unique series_title values.

In [17]:
series_df.nunique()
series_df['series_title'].unique()

series_id           8096
area_code             58
item_code            400
seasonal               2
periodicity_code       2
base_code              2
base_period           27
series_title        4229
footnote_codes         0
begin_year            65
begin_period          16
end_year              12
end_period            12
dtype: int64

array(['All items in U.S. city average, all urban consumers, seasonally adjusted',
       'Energy in U.S. city average, all urban consumers, seasonally adjusted',
       'All items less food in U.S. city average, all urban consumers, seasonally adjusted',
       ...,
       'Household furnishings and supplies in Portland-Salem, OR-WA, all urban consumers, not seasonally adjusted',
       'Transportation commodities less motor fuel in Portland-Salem, OR-WA, all urban consumers, not seasonally adjusted',
       'Gasoline, unleaded midgrade in Portland-Salem, OR-WA, all urban consumers, not seasonally adjusted'],
      dtype=object)

### The series_title values are of the form: '{item_name} in {area_name}, all urban consumers, {seasonal_text}'. 

In [18]:
series_df['series_title'][0]

'All items in U.S. city average, all urban consumers, seasonally adjusted'

In [19]:
series_df[series_df['series_title'].str.contains('all items',case=False,na=False)]

series_id area_code item_code seasonal periodicity_code  \
0     CUSR0000SA0            0000       SA0        S                R   
2     CUSR0000SA0L1          0000     SA0L1        S                R   
3     CUSR0000SA0L12         0000    SA0L12        S                R   
4     CUSR0000SA0L12E        0000   SA0L12E        S                R   
5     CUSR0000SA0L12E4       0000  SA0L12E4        S                R   
...                 ...       ...       ...      ...              ...   
8035  CUUSS49GSA0            S49G       SA0        U                S   
8037  CUUSS49GSA0L1E         S49G    SA0L1E        U                S   
8038  CUUSS49GSA0L2          S49G     SA0L2        U                S   
8039  CUUSS49GSA0L5          S49G     SA0L5        U                S   
8040  CUUSS49GSA0LE          S49G     SA0LE        U                S   

     base_code  base_period  \
0            S  1982-84=100   
2            S  1982-84=100   
3            S  1982-84=100   
4            S  1982-84=100   
5            S  1982-84=100   
...        ...          ...   
8035         S  1982-84=100   
8037         S  1982-84=100   
8038         S  1982-84=100   
8039         S  1982-84=100   
8040         S  1982-84=100   

                                           series_title  footnote_codes  \
0     All items in U.S. city average, all urban cons...             NaN   
2     All items less food in U.S. city average, all ...             NaN   
3     All items less food and shelter in U.S. city a...             NaN   
4     All items less food, shelter, and energy in U....             NaN   
5     All items less food, shelter, energy, and used...             NaN   
...                                                 ...             ...   
8035  All items in Urban Alaska, all urban consumers...             NaN   
8037  All items less food and energy in Urban Alaska...             NaN   
8038  All items less shelter in Urban Alaska, all ur...             NaN   
8039  All items less medical care in Urban Alaska, a...             NaN   
8040  All items less energy in Urban Alaska, all urb...             NaN   

      begin_year begin_period  end_year end_period  
0           1947          M01      2022        M06  
2           1947          M01      2022        M06  
3           1967          M01      2022        M06  
4           1967          M01      2022        M06  
5           1967          M01      2022        M06  
...          ...          ...       ...        ...  
8035        1984          S01      2022        S01  
8037        1984          S01      2022        S01  
8038        1984          S01      2022        S01  
8039        1984          S01      2022        S01  
8040        1984          S01      2022        S01  

[695 rows x 13 columns]

### The various types of items that we are looking at.

In [20]:
items = []
for string in item_df['item_name'].unique():
    items.append(string.split(',')[0])
set(items)

{'Admission to movies',
 'Admission to sporting events',
 'Admissions',
 'Airline fares',
 'Alcoholic beverages',
 'Alcoholic beverages at home',
 'Alcoholic beverages away from home',
 'All items',
 'All items - old base',
 'All items less energy',
 'All items less food',
 'All items less food and energy',
 'All items less food and shelter',
 'All items less medical care',
 'All items less shelter',
 'Apparel',
 'Apparel less footwear',
 'Apparel services other than laundry and dry cleaning',
 'Apples',
 'Appliances',
 'Audio equipment',
 'Automobile service clubs',
 'Baby food',
 'Bacon',
 'Bacon and related products',
 'Bakery products',
 'Bananas',
 'Bedroom furniture',
 'Beef and veal',
 'Beer',
 'Beverage materials including coffee and tea',
 "Boys' and girls' footwear",
 "Boys' apparel",
 'Bread',
 'Bread other than white',
 'Breakfast cereal',
 'Breakfast sausage and related products',
 'Butter',
 'Butter and margarine',
 'Cable and satellite television service',
 'Cakes',
 'Ca

### 7778 items are not seasonally adjusted; 318 items are seasonally adjusted.

In [21]:
series_df['seasonal'].unique()
series_df.groupby('seasonal').count()

array(['S', 'U'], dtype=object)

series_id  area_code  item_code  periodicity_code  base_code  \
seasonal                                                                 
S               318        318        318               318        318   
U              7778       7778       7778              7778       7778   

          base_period  series_title  footnote_codes  begin_year  begin_period  \
seasonal                                                                        
S                 318           318               0         318           318   
U                7778          7778               0        7778          7778   

          end_year  end_period  
seasonal                        
S              318         318  
U             7778        7778

### There doesn't seem to be a clear trend regarding what types of items are calculated seasonally or not.

In [22]:
set(series_df[series_df['seasonal'] == 'S']['series_title'])

{'Admission to movies, theaters, and concerts in U.S. city average, all urban consumers, seasonally adjusted',
 'Admission to sporting events in U.S. city average, all urban consumers, seasonally adjusted',
 'Admissions in U.S. city average, all urban consumers, seasonally adjusted',
 'Airline fares in U.S. city average, all urban consumers, seasonally adjusted',
 'Alcoholic beverages at home in U.S. city average, all urban consumers, seasonally adjusted',
 'Alcoholic beverages away from home in U.S. city average, all urban consumers, seasonally adjusted',
 'Alcoholic beverages in U.S. city average, all urban consumers, seasonally adjusted',
 'All items  less medical care in U.S. city average, all urban consumers, seasonally adjusted',
 'All items in U.S. city average, all urban consumers, seasonally adjusted',
 'All items less energy in U.S. city average, all urban consumers, seasonally adjusted',
 'All items less food and energy in U.S. city average, all urban consumers, seasonally a

In [23]:
set(series_df[series_df['seasonal'] == 'U']['series_title'])

{'Energy services in Midwest - Size Class D, all urban consumers, not seasonally adjusted',
 'Commodities in Washington-Baltimore, DC-MD-VA-WV, all urban consumers, not seasonally adjusted',
 'Toys, games, hobbies and playground equipment in U.S. city average, all urban consumers, not seasonally adjusted',
 'Alcoholic beverages in Mountain, all urban consumers, not seasonally adjusted',
 'Transportation in Riverside-San Bernardino-Ontario, CA, all urban consumers, not seasonally adjusted',
 'Motor vehicle insurance in Riverside-San Bernardino-Ontario, CA, all urban consumers, not seasonally adjusted',
 'Apparel in South - Size Class D, all urban consumers, not seasonally adjusted',
 'All items in Size Class B/C, all urban consumers, not seasonally adjusted',
 'Gasoline, unleaded regular in Pacific, all urban consumers, not seasonally adjusted',
 'Dairy and related products in Northeast - Size Class B/C, all urban consumers, not seasonally adjusted',
 'Commodities in U.S. city average, 

### Let's check the first and last records in the series_df, and see if they exist in the item_df and area_df ( i.e. the series_title contains the item_name from item_df and area_name from area_df ).

In [24]:
series_df[['series_id','series_title','item_code','area_code']].head(1)
item_df[item_df['item_code']==series_df['item_code'][0]]
area_df[area_df['area_code']==series_df['area_code'][0]]

series_id                                       series_title  \
0  CUSR0000SA0        All items in U.S. city average, all urban cons...   

  item_code area_code  
0       SA0      0000

item_code  item_name  display_level selectable  sort_sequence
2       SA0  All items              0          T              1

area_code          area_name  display_level selectable  sort_sequence
0      0000  U.S. city average              0          T              1

In [25]:
series_df[['series_id','series_title','item_code','area_code']].tail(1)
item_df[item_df['item_code']==series_df['item_code'][8095]]
area_df[area_df['area_code']==series_df['area_code'][8095]]

series_id                                       series_title  \
8095  CUUSS49GSS47016    Gasoline, unleaded premium in Urban Alaska, al...   

     item_code area_code  
8095   SS47016      S49G

item_code                   item_name  display_level selectable  \
370   SS47016  Gasoline, unleaded premium              4          T   

     sort_sequence  
370            225

area_code     area_name  display_level selectable  sort_sequence
57      S49G  Urban Alaska              1          T             44

### Because the number of uniques in the area_df and item_df is the same as the number of uniques in the area_code and item_code columns of the series_df, we could assume that there is a one-to-one correspondence of these values ( not necessarily true; e.g. the lists [1,2] and [1,3] both contain 2 unique values, but they do not share all unique values ). Are all the codes in area_df and item_df in series_df? We can verify this by the Boolean operations below:

In [26]:
# Are the area code values one-to-one?
set(series_df['area_code'].unique()) == set(area_df['area_code'])

# Are the item code values one-to-one?
set(series_df['item_code'].unique()) == set(item_df['item_code'])

True

True

### There are 434 series_id values in the series_df that are not in the cpi_df data. Let's drop these values before further analyses.

In [27]:
# Identify unique series_id values in each DataFrame.
cpiUniqueIDs = cpi_df['series_id'].unique()
seriesUniqueIDs = series_df['series_id'].unique()
len(cpiUniqueIDs)
len(seriesUniqueIDs)

# Create list of shared series_ids, and drop those not shared from series_df.
sharedIDs = list(set(cpiUniqueIDs).intersection(seriesUniqueIDs))
series_df = series_df[series_df['series_id'].isin(sharedIDs)]

# Verify that sets of series_id values match.
set(cpi_df['series_id'].unique()) == set(series_df['series_id'].unique())

7928

8096

True

### Let's verify the periodicity_code, base_code, base_period, begin_year, begin_period, end_year, and end_period are already contained in the cpi_df data.

In [28]:
series_df['begin_year'].min()
cpi_df['year'].min()

series_df['end_year'].max()
cpi_df['year'].max()

1913

1997

2022

2022

### Unfortunately, it looks like not even the begin_year values match up, so let's just drop these columns from series_df, since there relevance to the actual data is questionable.

### Let's drop all columns in series_df besides series_id, area_code, item_code, and seasonal.

In [29]:
series_df.columns.tolist()

['series_id',
 'area_code',
 'item_code',
 'seasonal',
 'periodicity_code',
 'base_code',
 'base_period',
 'series_title',
 'footnote_codes',
 'begin_year',
 'begin_period',
 'end_year',
 'end_period']

In [30]:
series_df.head()
series_df = series_df.drop(labels=['periodicity_code',
                       'base_code',
                       'base_period',
                       'series_title',
                       'footnote_codes',
                       'begin_year',
                       'begin_period',
                       'end_year',
                       'end_period'],axis=1)
series_df.head()

series_id area_code item_code seasonal periodicity_code base_code  \
0  CUSR0000SA0            0000       SA0        S                R         S   
1  CUSR0000SA0E           0000      SA0E        S                R         S   
2  CUSR0000SA0L1          0000     SA0L1        S                R         S   
3  CUSR0000SA0L12         0000    SA0L12        S                R         S   
4  CUSR0000SA0L12E        0000   SA0L12E        S                R         S   

   base_period                                       series_title  \
0  1982-84=100  All items in U.S. city average, all urban cons...   
1  1982-84=100  Energy in U.S. city average, all urban consume...   
2  1982-84=100  All items less food in U.S. city average, all ...   
3  1982-84=100  All items less food and shelter in U.S. city a...   
4  1982-84=100  All items less food, shelter, and energy in U....   

   footnote_codes  begin_year begin_period  end_year end_period  
0             NaN        1947          M01      2022        M06  
1             NaN        1957          M01      2022        M06  
2             NaN        1947          M01      2022        M06  
3             NaN        1967          M01      2022        M06  
4             NaN        1967          M01      2022        M06

series_id area_code item_code seasonal
0  CUSR0000SA0            0000       SA0        S
1  CUSR0000SA0E           0000      SA0E        S
2  CUSR0000SA0L1          0000     SA0L1        S
3  CUSR0000SA0L12         0000    SA0L12        S
4  CUSR0000SA0L12E        0000   SA0L12E        S

## df

### Let's create a new DataFrame df that merges the series_df, cpi_df, area_df, item_df, and seasonal_df features.

### First, let's merge the series_df with the area_code, item_code, and seasonal_text values from their respective DataFrames.

### Merge in area_name, item_name, and seasonal_text features.

In [31]:
series_df.head()
series_df = pd.merge(series_df,area_df.set_index('area_code')['area_name'],on='area_code')
series_df = pd.merge(series_df,item_df.set_index('item_code')['item_name'],on='item_code')
series_df = series_df.rename(columns = {'seasonal':'seasonal_code'})
series_df = pd.merge(series_df,seasonal_df.set_index('seasonal_code')['seasonal_text'],on='seasonal_code')
series_df.head()

series_id area_code item_code seasonal
0  CUSR0000SA0            0000       SA0        S
1  CUSR0000SA0E           0000      SA0E        S
2  CUSR0000SA0L1          0000     SA0L1        S
3  CUSR0000SA0L12         0000    SA0L12        S
4  CUSR0000SA0L12E        0000   SA0L12E        S

series_id area_code item_code seasonal_code          area_name  \
0  CUSR0000SA0            0000       SA0             S  U.S. city average   
1  CUSR0000SA0E           0000      SA0E             S  U.S. city average   
2  CUSR0000SA0L1          0000     SA0L1             S  U.S. city average   
3  CUSR0000SA0L12         0000    SA0L12             S  U.S. city average   
4  CUSR0000SA0L12E        0000   SA0L12E             S  U.S. city average   

                                  item_name        seasonal_text  
0                                 All items  Seasonally Adjusted  
1                                    Energy  Seasonally Adjusted  
2                       All items less food  Seasonally Adjusted  
3           All items less food and shelter  Seasonally Adjusted  
4  All items less food, shelter, and energy  Seasonally Adjusted

### Now we can drop the area_code, item_code, and seasonal_code features.

In [32]:
series_df.head()
series_df = series_df.drop(labels=['area_code','item_code','seasonal_code'],axis=1)
series_df.head()

series_id area_code item_code seasonal_code          area_name  \
0  CUSR0000SA0            0000       SA0             S  U.S. city average   
1  CUSR0000SA0E           0000      SA0E             S  U.S. city average   
2  CUSR0000SA0L1          0000     SA0L1             S  U.S. city average   
3  CUSR0000SA0L12         0000    SA0L12             S  U.S. city average   
4  CUSR0000SA0L12E        0000   SA0L12E             S  U.S. city average   

                                  item_name        seasonal_text  
0                                 All items  Seasonally Adjusted  
1                                    Energy  Seasonally Adjusted  
2                       All items less food  Seasonally Adjusted  
3           All items less food and shelter  Seasonally Adjusted  
4  All items less food, shelter, and energy  Seasonally Adjusted

series_id          area_name  \
0  CUSR0000SA0        U.S. city average   
1  CUSR0000SA0E       U.S. city average   
2  CUSR0000SA0L1      U.S. city average   
3  CUSR0000SA0L12     U.S. city average   
4  CUSR0000SA0L12E    U.S. city average   

                                  item_name        seasonal_text  
0                                 All items  Seasonally Adjusted  
1                                    Energy  Seasonally Adjusted  
2                       All items less food  Seasonally Adjusted  
3           All items less food and shelter  Seasonally Adjusted  
4  All items less food, shelter, and energy  Seasonally Adjusted

### Next, let's merge some features into the cpi_df.

### First, let's drop the average period values.

In [33]:
cpi_df['period'].unique()
cpi_df = cpi_df[cpi_df['period'].isin(['M13','S01','S02','S03']) == False]

array(['M01', 'M02', 'M03', 'M04', 'M05', 'M06', 'M07', 'M08', 'M09',
       'M10', 'M11', 'M12', 'M13', 'S01', 'S02', 'S03'], dtype=object)

In [34]:
# Sum is zero if values are not in period feature.
sum(cpi_df['period'].isin(['M13','S01','S02','S03']))

0

### Next, let's merge in the period_name feature from the period_df.

In [35]:
cpi_df.head()
cpi_df = pd.merge(cpi_df,period_df.set_index('period')['period_name'],on='period')
cpi_df.head()

series_id  year period  value
0  CUSR0000SA0        1997    M01  159.4
1  CUSR0000SA0        1997    M02  159.7
2  CUSR0000SA0        1997    M03  159.8
3  CUSR0000SA0        1997    M04  159.9
4  CUSR0000SA0        1997    M05  159.9

series_id  year period  value period_name
0  CUSR0000SA0        1997    M01  159.4     January
1  CUSR0000SA0        1998    M01  162.0     January
2  CUSR0000SA0        1999    M01  164.7     January
3  CUSR0000SA0        2000    M01  169.3     January
4  CUSR0000SA0        2001    M01  175.6     January

### Next, let's convert the year and period_name into a datetime object feature.

In [36]:
cpi_df.head()
cpi_df['date'] = pd.to_datetime(cpi_df['year'].apply(str) + cpi_df['period_name'],format='%Y%B')
cpi_df.head()

series_id  year period  value period_name
0  CUSR0000SA0        1997    M01  159.4     January
1  CUSR0000SA0        1998    M01  162.0     January
2  CUSR0000SA0        1999    M01  164.7     January
3  CUSR0000SA0        2000    M01  169.3     January
4  CUSR0000SA0        2001    M01  175.6     January

series_id  year period  value period_name       date
0  CUSR0000SA0        1997    M01  159.4     January 1997-01-01
1  CUSR0000SA0        1998    M01  162.0     January 1998-01-01
2  CUSR0000SA0        1999    M01  164.7     January 1999-01-01
3  CUSR0000SA0        2000    M01  169.3     January 2000-01-01
4  CUSR0000SA0        2001    M01  175.6     January 2001-01-01

### We can drop the year, period, and period_name features now.

In [37]:
cpi_df.head()
cpi_df = cpi_df.drop(['year','period','period_name'],axis=1)
cpi_df.head()

series_id  year period  value period_name       date
0  CUSR0000SA0        1997    M01  159.4     January 1997-01-01
1  CUSR0000SA0        1998    M01  162.0     January 1998-01-01
2  CUSR0000SA0        1999    M01  164.7     January 1999-01-01
3  CUSR0000SA0        2000    M01  169.3     January 2000-01-01
4  CUSR0000SA0        2001    M01  175.6     January 2001-01-01

series_id  value       date
0  CUSR0000SA0        159.4 1997-01-01
1  CUSR0000SA0        162.0 1998-01-01
2  CUSR0000SA0        164.7 1999-01-01
3  CUSR0000SA0        169.3 2000-01-01
4  CUSR0000SA0        175.6 2001-01-01

### Now, let's merge the cpi_df and series_df into one DataFrame df.

In [38]:
df = pd.merge(cpi_df,series_df,on='series_id')
df.head()

series_id  value       date          area_name  item_name  \
0  CUSR0000SA0        159.4 1997-01-01  U.S. city average  All items   
1  CUSR0000SA0        162.0 1998-01-01  U.S. city average  All items   
2  CUSR0000SA0        164.7 1999-01-01  U.S. city average  All items   
3  CUSR0000SA0        169.3 2000-01-01  U.S. city average  All items   
4  CUSR0000SA0        175.6 2001-01-01  U.S. city average  All items   

         seasonal_text  
0  Seasonally Adjusted  
1  Seasonally Adjusted  
2  Seasonally Adjusted  
3  Seasonally Adjusted  
4  Seasonally Adjusted

### Let's drop all of the Seasonally Adjusted data, and then drop the seasonal_text column.

In [39]:
df = df[df['seasonal_text']=='Seasonally Adjusted']
df['seasonal_text'].unique().tolist()
df = df.drop(['seasonal_text'],axis=1)
df.head()

['Seasonally Adjusted']

series_id  value       date          area_name  item_name
0  CUSR0000SA0        159.4 1997-01-01  U.S. city average  All items
1  CUSR0000SA0        162.0 1998-01-01  U.S. city average  All items
2  CUSR0000SA0        164.7 1999-01-01  U.S. city average  All items
3  CUSR0000SA0        169.3 2000-01-01  U.S. city average  All items
4  CUSR0000SA0        175.6 2001-01-01  U.S. city average  All items

### We can now drop the series_id feature, and, for visual convenience, we can rename and reorder the columns ( columns ordered from left to right in increasing specificity ).

In [40]:
df.head()
df = df.drop(['series_id'],axis=1)
df = df.rename(columns={'area_name':'area','item_name':'item'})
df = df[['area','date','item','value']]
df.head()

series_id  value       date          area_name  item_name
0  CUSR0000SA0        159.4 1997-01-01  U.S. city average  All items
1  CUSR0000SA0        162.0 1998-01-01  U.S. city average  All items
2  CUSR0000SA0        164.7 1999-01-01  U.S. city average  All items
3  CUSR0000SA0        169.3 2000-01-01  U.S. city average  All items
4  CUSR0000SA0        175.6 2001-01-01  U.S. city average  All items

area       date       item  value
0  U.S. city average 1997-01-01  All items  159.4
1  U.S. city average 1998-01-01  All items  162.0
2  U.S. city average 1999-01-01  All items  164.7
3  U.S. city average 2000-01-01  All items  169.3
4  U.S. city average 2001-01-01  All items  175.6

# Visualizations

### Let's get an idea of the various items and areas we are working with.

In [41]:
df.info()
df.describe().transpose()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88694 entries, 0 to 88693
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   area    88694 non-null  object        
 1   date    88694 non-null  datetime64[ns]
 2   item    88694 non-null  object        
 3   value   88694 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 3.4+ MB


count        mean         std    min    25%    50%        75%  \
value  88694.0  192.929021  108.811469  1.269  123.4  170.1  233.14775   

            max  
value  1346.046

In [42]:
df.columns

Index(['area', 'date', 'item', 'value'], dtype='object')

In [43]:
df['area'].nunique()
df['area'].unique().tolist()

1

['U.S. city average']

In [44]:
df['item'].nunique()
df['item'].unique().tolist()

318

['All items',
 'Energy',
 'All items less food',
 'All items less food and shelter',
 'All items less food, shelter, and energy',
 'All items less food, shelter, energy, and used cars and trucks',
 'All items less food and energy',
 'All items less shelter',
 'All items less medical care',
 'All items less energy',
 'Apparel less footwear',
 'Apparel',
 "Men's and boys' apparel",
 "Women's and girls' apparel",
 'Commodities',
 'Energy commodities',
 'Commodities less food',
 'Commodities less food and beverages',
 'Commodities less food and energy commodities',
 'Commodities less food, energy, and used cars and trucks',
 'Durables',
 'Education and communication',
 'Education',
 'Communication',
 'Information and information processing',
 'Education and communication commodities',
 'Education and communication services',
 'Food and beverages',
 'Food',
 'Food at home',
 'Cereals and bakery products',
 'Meats, poultry, fish, and eggs',
 'Meats, poultry, and fish',
 'Meats',
 'Fruits and

### Let's create lists for the various types of locations.

In [45]:
cities = ['U.S. city average',
          'Pittsburgh, PA',
          'Cleveland-Akron, OH',
          'Washington-Baltimore, DC-MD-VA-WV',
          'Los Angeles-Riverside-Orange County, CA',
          'Boston-Cambridge-Newton, MA-NH',
          'New York-Newark-Jersey City, NY-NJ-PA',
          'Philadelphia-Camden-Wilmington, PA-NJ-DE-MD',
          'Chicago-Naperville-Elgin, IL-IN-WI',
          'Detroit-Warren-Dearborn, MI',
          'Minneapolis-St.Paul-Bloomington, MN-WI',
          'St. Louis, MO-IL',
          'Washington-Arlington-Alexandria, DC-VA-MD-WV',
          'Miami-Fort Lauderdale-West Palm Beach, FL',
          'Atlanta-Sandy Springs-Roswell, GA',
          'Tampa-St. Petersburg-Clearwater, FL',
          'Baltimore-Columbia-Towson, MD',
          'Dallas-Fort Worth-Arlington, TX',
          'Houston-The Woodlands-Sugar Land, TX',
          'Phoenix-Mesa-Scottsdale, AZ',
          'Denver-Aurora-Lakewood, CO',
          'Los Angeles-Long Beach-Anaheim, CA',
          'San Francisco-Oakland-Hayward, CA',
          'Riverside-San Bernardino-Ontario, CA',
          'Seattle-Tacoma-Bellevue WA',
          'San Diego-Carlsbad, CA',
          'Urban Hawaii',
          'Urban Alaska',
          'Milwaukee-Racine, WI',
          'Cincinnati-Hamilton, OH-KY-IN',
          'Kansas City, MO-KS',
          'Portland-Salem, OR-WA']

regions = ['Northeast',
           'New England',
           'Middle Atlantic',
           'Midwest',
           'East North Central',
           'West North Central',
           'South',
           'South Atlantic',
           'East South Central',
           'West South Central',
           'West',
           'Mountain',
           'Pacific']

classes = ['Size Class D',
           'Midwest - Size Class D',
           'South - Size Class D',
           'Size Class B/C',
           'Northeast - Size Class B/C',
           'Midwest - Size Class B/C',
           'South - Size Class B/C',
           'West - Size Class B/C',
           'Size Class A',
           'Northeast - Size Class A',
           'Midwest - Size Class A',
           'South - Size Class A']

### Let's generate a DataFrame containing data for the U.S. city average CPI value of all items, not seasonally adjusted.

In [46]:
all_items_df = df[(df['item'] == 'All items') 
           & (df['area'] == 'U.S. city average')].sort_values(by='date')

# Add monthly and yearly changes in value.
all_items_df['monthly'] = 100*all_items_df['value'].pct_change()
all_items_df['yearly'] = 100*all_items_df['value'].pct_change(periods=12)

# Drop irrelevant columns.
all_items_df = all_items_df.drop(['area','item'],axis=1)
all_items_df.head()

date  value   monthly  yearly
0   1997-01-01  159.4       NaN     NaN
26  1997-02-01  159.7  0.188206     NaN
52  1997-03-01  159.8  0.062617     NaN
78  1997-04-01  159.9  0.062578     NaN
104 1997-05-01  159.9  0.000000     NaN

In [47]:
all_items_df.iplot(kind='scatter',x='date',y='value',
                 title='U.S. City Average Value, All Items, Not Seasonally Adjusted',
                 xTitle='Date',
                 yTitle='CPI Value',
                 color='black')

In [48]:
all_items_df.iplot(kind='scatter',x='date',y='monthly',
                 title='MoM Change in U.S. City Average Value, All Items, Not Seasonally Adjusted',
                 xTitle='Date',
                 yTitle='MoM Change in CPI Value (%)',
                 color='black')

In [49]:
all_items_df.iplot(kind='scatter',x='date',y='yearly',
                 title='YoY Change in U.S. City Average Value, All Items, Not Seasonally Adjusted',
                 xTitle='Date',
                 yTitle='YoY Change in CPI Value (%)',`
                 color='black')

SyntaxError: invalid syntax (3290890479.py, line 4)